File: app/src/main/java/org/fireking/basic/recyclerview/WanNianLiHomeActivity.java
package org.fireking.basic.recyclerview;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager.widget.ViewPager;

import net.lucode.hackware.magicindicator.MagicIndicator;
import net.lucode.hackware.magicindicator.ViewPagerHelper;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.CommonNavigator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.CommonNavigatorAdapter;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerIndicator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerTitleView;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.indicators.LinePagerIndicator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.titles.ColorTransitionPagerTitleView;

import org.fireking.ap.R;

import java.util.ArrayList;
import java.util.List;


public class WanNianLiHomeActivity extends AppCompatActivity {

    public static void start(Context context) {
        context.startActivity(new Intent(context, WanNianLiHomeActivity.class));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_wan_nian_li_home);

        final List<String> titles = new ArrayList<>();
        titles.add("推荐");
        titles.add("运势");
        titles.add("视频");
        titles.add("健康");
        titles.add("星座");
        titles.add("笑话");
        titles.add("每日一言");
        titles.add("今日星座");
        titles.add("生肖");

        final ViewPager newsViewPager = findViewById(R.id.newsViewPager);
        newsViewPager.setAdapter(new WanNianliHomeFragmentAdapter(getSupportFragmentManager(), titles));

        MagicIndicator magicIndicator = findViewById(R.id.magic_indicator);
        CommonNavigator commonNavigator = new CommonNavigator(this);
        commonNavigator.setAdapter(new CommonNavigatorAdapter() {

            @Override
            public int getCount() {
                return titles.size();
            }

            @Override
            public IPagerTitleView getTitleView(Context context, final int index) {
                ColorTransitionPagerTitleView colorTransitionPagerTitleView = new ColorTransitionPagerTitleView(context);
                colorTransitionPagerTitleView.setNormalColor(Color.GRAY);
                colorTransitionPagerTitleView.setSelectedColor(Color.BLACK);
                colorTransitionPagerTitleView.setText(titles.get(index));
                colorTransitionPagerTitleView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        newsViewPager.setCurrentItem(index);
                    }
                });
                return colorTransitionPagerTitleView;
            }

            @Override
            public IPagerIndicator getIndicator(Context context) {
                LinePagerIndicator indicator = new LinePagerIndicator(context);
                indicator.setMode(LinePagerIndicator.MODE_WRAP_CONTENT);
                return indicator;
            }
        });
        magicIndicator.setNavigator(commonNavigator);

        ViewPagerHelper.bind(magicIndicator, newsViewPager);
    }
}

File: app/src/main/java/org/fireking/ap/custom/recyclerview/WanNianLiHomeActivity.kt
package org.fireking.ap.custom.recyclerview

import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.android.vlayout.DelegateAdapter
import com.alibaba.android.vlayout.VirtualLayoutManager
import kotlinx.android.synthetic.main.activity_wan_nian_li_home.*
import org.fireking.ap.R
import org.fireking.ap.custom.recyclerview.adapter.WanNianLiBodyAdapter
import org.fireking.ap.custom.recyclerview.adapter.WanNianLiHeaderAdapter
import org.jetbrains.anko.intentFor

class WanNianLiHomeActivity : AppCompatActivity() {

    private lateinit var delegateAdapter: DelegateAdapter

    companion object {
        @JvmStatic
        fun start(context: Context) {
            context.startActivity(context.intentFor<WanNianLiHomeActivity>())
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_wan_nian_li_home)

        val virtualLayoutManager = VirtualLayoutManager(this)
        outRecyclerView.layoutManager = virtualLayoutManager
        val viewPool = RecyclerView.RecycledViewPool()
        viewPool.setMaxRecycledViews(0, 10)

        delegateAdapter = DelegateAdapter(virtualLayoutManager)
        outRecyclerView.adapter = delegateAdapter

        val headerAdapter = WanNianLiHeaderAdapter()
        delegateAdapter.addAdapter(headerAdapter)

        val bodyAdapter = WanNianLiBodyAdapter()
        delegateAdapter.addAdapter(bodyAdapter)
    }
}
